lib/sqlalchemy/sql/elements.py

```diff
@@ -3066,7 +3125,9 @@ def self_group(self, against=None):
 or_ = BooleanClauseList.or_
```

Comment:
Here method assigned to an attribute variable makes mypy complain - possibly because `BooleanClauseList` has a private constructor `_construct` while `__init__` simply raises a `NotImplementedError`. In mypy's error message - `"Callable[[VarArg(Any), KwArg(Any)], NoReturn]" has no attribute "and_"` - the `Callable`'s signature is the signature of `__init__` method. Not sure how to proceed with this

Reply:
I am wrong, `BooleanClauseList.and_` & `BooleanClauseList.or_` are classmethods - so python shouldn't need to call `BooleanClauseList.__init__`. Anyway, that doesn't explain where the `Callable` signature in the mypy error message came from

Reply:
Here's a minimal reproducible example:
```py
from typing import Any, NoReturn


class Foo:
    def __init__(self, *arg: Any, **kw: Any) -> NoReturn:
        raise NotImplementedError(
            "BooleanClauseList has a private constructor"
        )

    @classmethod
    def bar(cls) -> None:
        ...


bar = Foo.bar

```

Note that if you remove `__init__` (or make it return) `mypy` will stop complaining. Try to google around that.

Reply:
Btw, you can try something like that if you find no better way on Google:
```py
class Foo:
    if not TYPE_CHECKING:
        def __init__(self, *arg: Any, **kw: Any) -> NoReturn:
            raise NotImplementedError(
                "BooleanClauseList has a private constructor"
            )
```

Reply:
Thanks for the examples! Googling didn't help much, but I have found a comment from Mike in Dmitry's [PR](https://github.com/sqlalchemy/sqlalchemy/pull/9028) for `sql.selectable` regarding annotating the `__init__`: in that particular case he suggests removing its return annotation at all and leaving it implicit. Did that here, the error is gone, `_and` assigned correctly, and mypy doesn't complain about `__init__` having no return type. Will still leave a comment for the maintainers, just to make sure that Mike's advice wasn't regarding only that particular case in `sql.selectable`